Add base class and one implementation of a record filter

cmcantalupo opened the pull request:
- Add doxygen comments to filter classes
- Implemented simple epoch inference with the
  ProxyEpochRecordFilter
- Filtering done in the ApplicationSampler
- Add plumbing to add record filter to ApplicationSampler

Change-Id: Id5bc5b97d3cc2d5f5953d3580f6f403bb59452ce
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add base class and one implementation of a record filter

- Add doxygen comments to filter classes
- Implemented simple epoch inference with the
  ProxyEpochRecordFilter
- Filtering done in the ApplicationSampler
- Add plumbing to add record filter to ApplicationSampler

Change-Id: Id5bc5b97d3cc2d5f5953d3580f6f403bb59452ce
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dianarg commented on src/ProxyEpochRecordFilter.hpp:
```diff
@@ -0,0 +1,99 @@
+namespace geopm
+{
+    /// @brief Filter that can be used in place of the
+    ///        EpochRecordFilter to synthesize epoch events from a
```
EpochRecordFilter is going away, right?

dianarg commented on src/ProxyEpochRecordFilter.hpp:
```diff
@@ -0,0 +1,99 @@
+    /// @brief Filter that can be used in place of the
+    ///        EpochRecordFilter to synthesize epoch events from a
+    ///        sequence of region entry events.  The filter also
+    ///        passes hint events through.
```
update to indicate all events except epoch are passed though

dianarg commented on src/RecordFilter.cpp:
```diff
@@ -0,0 +1,79 @@
+
+namespace geopm
+{
+    std::unique_ptr<RecordFilter> RecordFilter::make_unique(const std::string &name)
```
please add tests for this

dianarg commented on src/ApplicationSampler.cpp:
```diff
@@ -86,7 +92,15 @@ namespace geopm
+
+    }
+
+    ApplicationSamplerImp::ApplicationSamplerImp(const std::string &filter_name)
```
it might be easier to test the passthrough if RecordFilters are injected instead of the name

cmcantalupo replied:
We need a way to generate record filters for each process after construction.  That is facilitated with a string now.  I'm not sure how we want to mock out the RecordFilter::make_unique() call, which is where we will run into testing problems.

cmcantalupo replied:
maybe:

```
ApplicationSamplerImp(std::function<std::unique_ptr<RecordFilter>()> filter_factory);
```

cmcantalupo committed:
Address review comments

- ApplicationSampler test constructor takes a RecordFilter
  factory not a name.
- Adds tests for RecordFilter factory.
- Clean up documentation for ProxyEpochRecordFilter.

Change-Id: I13aca63e533bb9d90afbb1296ad91cd66d02c203
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dianarg commented on test/RecordFilterTest.cpp:
```diff
@@ -0,0 +1,95 @@
+TEST_F(RecordFilterTest, make_proxy_epoch)
+{
+    std::shared_ptr<RecordFilter> filter = RecordFilter::make_unique("proxy_epoch,0xabcd1234");
+    EXPECT_TRUE(filter);
```
maybe EXPECT_NO_THROW would be clearer?

cmcantalupo committed:
Add doxygen comment for RecordFilter::parse_name_proxy_epoch()

Change-Id: Ifb58bed8b4ede2e48959fbbb4942884f068f3c40
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add more unit tests to cover changes

- Try out the filter returned by the RecordFilter factory.
- Add checks for the new Environment class APIs.

Change-Id: If26faa8615fd1297b31bc8ac3c279d66c23e112b
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
